src/model/Position.ts

```diff
@@ -40,6 +44,34 @@ export class Position {
     feeGrowthInside0X128: JSBI,
     feeGrowthInside1X128: JSBI
   ) {
-    // TODO
+    let liquidityNext: JSBI;
+    if (JSBI.equal(liquidityDelta, ZERO)) {
+      assert(JSBI.lessThanOrEqual(this.liquidity, ZERO), "NP");
+      liquidityNext = this.liquidity;
+    } else {
+      liquidityNext = LiquidityMath.addDelta(this.liquidity, liquidityDelta);
+    }
+
+    const tokensOwed0 = FullMath.mulDiv(
+      JSBI.subtract(feeGrowthInside0X128, this.feeGrowthInside0LastX128),
+      this.liquidity,
+      Q128
+    );
+    const tokensOwed1 = FullMath.mulDiv(
+      JSBI.subtract(feeGrowthInside1X128, this.feeGrowthInside1LastX128),
+      this.liquidity,
+      Q128
+    );
+
+    if (JSBI.notEqual(liquidityDelta, ZERO)) this._liquidity = liquidityNext;
+    this._feeGrowthInside0LastX128 = feeGrowthInside0X128;
+    this._feeGrowthInside1LastX128 = feeGrowthInside1X128;
+    if (
+      JSBI.greaterThan(tokensOwed0, ZERO) ||
+      JSBI.greaterThan(tokensOwed1, ZERO)
+    ) {
+      this._tokensOwed0 = JSBI.add(this.tokensOwed0, tokensOwed0);
+      this._tokensOwed1 = JSBI.add(this.tokensOwed1, tokensOwed1);
+    }
   }
```

Comment:
I realize we need another `update` method here and sure it's a breaking-change. Update above is to refresh fee income as well as adjust liquidity for both mint and burn, and the new one as `updateLiquidityBurn(amount0: JSBI, amount1: JSBI)` is to update tokensOwed from liquidity burn.

Reply:
Can you elaborate more on this? If the update already adjusts both mint and burn then why do we need an update just for liquidity burns? Is there code reference to this in uniswap-v3-core?

Reply:
Adjustment on `tokensOwed` is from two cases. They are fee income refreshment(mint and burn) and liquidity capital burn(burn only). `update` here only handles first case. In second case, code in uniswap-v3-core manupilate it directly which is not appropriate to our structure. It's difficult to add this logic to `updatePosition` as caller of `update` here because at that time amount0 and amount1 are still need to be calculated, which happens afterwards, in `modifyPosition` as caller of `updatePosition`. 
https://github.com/Uniswap/uniswap-v3-core/blob/b2c5555d696428c40c4b236069b3528b2317f3c1/contracts/UniswapV3Pool.sol#L536